nightly-examples/examples/echo.rs

```diff
@@ -0,0 +1,89 @@
+//! A "hello world" echo server with Tokio
+//!
+//! This server will create a TCP listener, accept connections in a loop, and
+//! write back everything that's read off of each TCP connection.
+//!
+//! Because the Tokio runtime uses a thread pool, each TCP connection is
+//! processed concurrently with all other TCP connections across multiple
+//! threads.
+//!
+//! To see this server in action, you can run this in one terminal:
+//!
+//!     cargo run --example echo
+//!
+//! and in another terminal you can run:
+//!
+//!     cargo run --example connect 127.0.0.1:8080
+//!
+//! Each line you type in to the `connect` terminal should be echo'd back to
+//! you! If you open up multiple terminals running the `connect` example you
+//! should be able to see them all make progress simultaneously.
+
+#![feature(async_await)]
+#![warn(rust_2018_idioms)]
+
+use tokio;
+use tokio::io::{AsyncReadExt, AsyncWriteExt};
+use tokio::net::TcpListener;
+
+use std::env;
+use std::error::Error;
+use std::net::SocketAddr;
+
+use tracing::{debug, error, info, warn};
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn Error>> {
+    let subscriber = tracing_fmt::FmtSubscriber::builder().finish();
+    tracing::subscriber::set_global_default(subscriber)?;
+
+    // Allow passing an address to listen on as the first argument of this
+    // program, but otherwise we'll just set up our TCP listener on
+    // 127.0.0.1:8080 for connections.
+    let addr = env::args().nth(1).unwrap_or("127.0.0.1:3000".to_string());
+    let addr = addr.parse::<SocketAddr>()?;
+
+    // Next up we create a TCP listener which will listen for incoming
+    // connections. This TCP listener is bound to the address we determined
+    // above and must be associated with an event loop.
+    let mut listener = TcpListener::bind(&addr)?;
+    info!("Listening on: {}", addr);
+
+    loop {
+        // Asynchronously wait for an inbound socket.
+        let (mut socket, peer_addr) = listener.accept().await?;
+
+        info!("Got connection from: {}", peer_addr);
+
+        // And this is where much of the magic of this server happens. We
+        // crucially want all clients to make progress concurrently, rather than
+        // blocking one on completion of another. To achieve this we use the
+        // `tokio::spawn` function to execute the work in the background.
+        //
+        // Essentially here we're executing a new task to run concurrently,
+        // which will allow all of our clients to be processed concurrently.
+
+        tokio::spawn(async move {
+            let mut buf = [0; 1024];
+
+            // In a loop, read data from the socket and write the data back.
+            loop {
+                let n = socket
+                    .read(&mut buf)
```

Comment:
WDYT about adding some additional tracing instrumentation here, like:
```rust
let n = socket
    .read(&mut buf)
    .then(|res| match res {
        Ok(bytes) => {
            trace!(bytes)
            Ok(bytes)
        },
        Err(error) => {
            warn!(%error)
            Err(error)
        },
    })
    .instrument(trace_span!("read"))
    .await
    .expect("failed to read data from socket");
```

We could also make this a bit simpler by just recording the read/error events outside the chained future, but i thought it was a neat demo of how we can instrument something with a span and then `await` it...

Reply:
Yeah, I think this is good example of better instrumentation. I wasn't entirely sure how the `.instrument` method would work with the `#[instrument]` macro, so I didn't add it anywhere in the examples.

Reply:
> I wasn't entirely sure how the `.instrument` method would work with the `#[instrument]` macro, so I didn't add it anywhere in the examples.

They should be compatible: the `#[instrument]` macro instruments a function, while `Future::instrument` instruments an expression implementing `Future`. :)